### PR TITLE
Added cursor palette support to GIMP/Jasc palette definitions

### PR DESCRIPTION
**Pull request by Mailaender**
Followup of https://github.com/OpenRA/OpenRA/pull/17538 which allows me to use the original palette without butchering the original cursor sprites. This may be a nice feature for other mods as well.

image

```
	PaletteFromGimpOrJascFile@cursor:
		Name: cursor
		Filename: palette.pal
		CursorPalette: true
		AllowModifiers: false
		TransparentIndex: 255
```

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/World/PaletteFromGimpOrJascFile.cs [hunk 1]
```diff
@@ -45,7 +46,68 @@ class PaletteFromGimpOrJascFileInfo : ITraitInfo
...
+		[Desc("Whether this palette is available for cursors.")]
+		public readonly bool CursorPalette = false;
+
 		public object Create(ActorInitializer init) { return new PaletteFromGimpOrJascFile(init.World, this); }
+
+		string IProvidesCursorPaletteInfo.Palette { get { return CursorPalette ? Name : null; } }
+
+		ImmutablePalette IProvidesCursorPaletteInfo.ReadPalette(IReadOnlyFileSystem fileSystem)
+		{
+			var s = fileSystem.Open(Filename);
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/World/PaletteFromGimpOrJascFile.cs]: Is there a reason you dropped the `using` from this?

Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/World/PaletteFromGimpOrJascFile.cs]: That was an accident. Added it back.

**Comment by pchote** [comment]
Note that once #16885 lands you won't need to use cursor palettes at all.

File: OpenRA.Mods.Common/Traits/World/PaletteFromGimpOrJascFile.cs [hunk 2]
```diff
@@ -45,32 +46,25 @@ class PaletteFromGimpOrJascFileInfo : ITraitInfo
...
-		public void LoadPalettes(WorldRenderer wr)
-		{
-			var colors = new uint[Palette.Size];
-			using (var s = world.Map.Open(info.Filename))
+			using (var s = fileSystem.Open(Filename))
 			{
+				var colors = new uint[Palette.Size];
 				using (var lines = s.ReadAllLines().GetEnumerator())
 				{
 					if (lines == null)
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/PaletteFromGimpOrJascFile.cs]: Lines can never be null, so this and the exception can be removed.

Screenshot 2020-01-12 at 19 41 27

Comment (Mailaender) [reply OpenRA.Mods.Common/Traits/World/PaletteFromGimpOrJascFile.cs]: Fixed

**Comment by Mailaender** [comment]
My mouse pointer are 8 bit artwork so I don't need that. If you need help testing this simply run https://github.com/Mailaender/OpenHV as I based the mod upon this last commit.

**Review by pchote — APPROVED** [review]

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/75f75c281e39f0ea073460e39d51387e3cf2d84d...c83f99cbdecbed94eea690b03a6532f967edcf79)

**Pull request merged**